osgp/platform/osgp-adapter-domain-smartmetering/src/main/java/org/opensmartgridplatform/adapter/domain/smartmetering/infra/jms/core/messageprocessors/ActualPowerQualityResponseMessageProcessor.java

```diff
@@ -0,0 +1,64 @@
+/**
+ * Copyright 2021 Alliander N.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.adapter.domain.smartmetering.infra.jms.core.messageprocessors;
+
+import org.opensmartgridplatform.adapter.domain.smartmetering.application.services.MonitoringService;
+import org.opensmartgridplatform.adapter.domain.smartmetering.infra.jms.core.OsgpCoreResponseMessageProcessor;
+import org.opensmartgridplatform.adapter.domain.smartmetering.infra.jms.ws.WebServiceResponseMessageSender;
+import org.opensmartgridplatform.dto.valueobjects.smartmetering.ActualPowerQualityResponseDto;
+import org.opensmartgridplatform.shared.exceptionhandling.ComponentType;
+import org.opensmartgridplatform.shared.exceptionhandling.FunctionalException;
+import org.opensmartgridplatform.shared.exceptionhandling.FunctionalExceptionType;
+import org.opensmartgridplatform.shared.exceptionhandling.OsgpException;
+import org.opensmartgridplatform.shared.infra.jms.DeviceMessageMetadata;
+import org.opensmartgridplatform.shared.infra.jms.MessageProcessorMap;
+import org.opensmartgridplatform.shared.infra.jms.MessageType;
+import org.opensmartgridplatform.shared.infra.jms.ResponseMessage;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.stereotype.Component;
+
+@Component
+public class ActualPowerQualityResponseMessageProcessor extends OsgpCoreResponseMessageProcessor {
+
+    @Autowired
+    @Qualifier("domainSmartMeteringMonitoringService")
+    private MonitoringService monitoringService;
+
+    @Autowired
+    protected ActualPowerQualityResponseMessageProcessor(
+            WebServiceResponseMessageSender responseMessageSender,
+            @Qualifier("domainSmartMeteringInboundOsgpCoreResponsesMessageProcessorMap") MessageProcessorMap messageProcessorMap) {
+        super(responseMessageSender, messageProcessorMap, MessageType.GET_ACTUAL_POWER_QUALITY,
+                ComponentType.DOMAIN_SMART_METERING);
+    }
+
+    @Override
+    protected boolean hasRegularResponseObject(final ResponseMessage responseMessage) {
+        final Object dataObject = responseMessage.getDataObject();
+        return dataObject instanceof ActualPowerQualityResponseDto;
+    }
+
+    @Override
+    protected void handleMessage(final DeviceMessageMetadata deviceMessageMetadata,
+            final ResponseMessage responseMessage, final OsgpException osgpException) throws FunctionalException {
+
+        if (hasRegularResponseObject(responseMessage)) {
```

Comment:
I would rewrite this code.

final ActualPowerQualityResponseDto actualPowerQualityResponseDto = (ActualPowerQualityResponseDto) responseMessage
                    .getDataObject();

in case of classcastexception you catch the error and throw the FunctionalException. 

this code will be much clearer and you don;t need the hasRegularResponseObject() method.

Reply:
Method hasRegularResponseObject() is required by the abstract superclass OsgpCoreResponseMessageProcessor. Is why not use it.